includes/class-wc-product-external.php

```diff
@@ -65,23 +146,37 @@ public function add_to_cart_text() {
 		return apply_filters( 'woocommerce_product_single_add_to_cart_text', $this->get_button_text(), $this );
 	}
 
+	/*
+	|--------------------------------------------------------------------------
+	| CRUD methods
+	|--------------------------------------------------------------------------
+	*/
+
 	/**
-	 * Get product url.
+	 * Reads a product from the database and sets its data to the class.
 	 *
-	 * @access public
-	 * @return string
+	 * @since 2.7.0
+	 * @param int $id Product ID.
 	 */
-	public function get_product_url() {
-		return esc_url( $this->product_url );
+	public function read( $id ) {
+		parent::read( $id );
+		$this->set_props( array(
+			'product_url' => get_post_meta( $id, '_product_url', true ),
+			'button_text' => get_post_meta( $id, '_button_text', true ),
+		) );
+		do_action( 'woocommerce_product_loaded', $this );
+		do_action( 'woocommerce_product_' . $this->get_type() . '_loaded', $this );
 	}
 
 	/**
-	 * Get button text.
+	 * Helper method that updates all the post meta for an external product.
 	 *
-	 * @access public
-	 * @return string
+	 * @since 2.7.0
+	 * @param int $id
 	 */
-	public function get_button_text() {
-		return $this->button_text ? $this->button_text : __( 'Buy product', 'woocommerce' );
+	protected function update_post_meta( $id ) {
```

Comment:
I've noticed in all the other CRUD classes, we accept an `$id` here. I can't think of any case where we would call update_post_meta and pass a different ID. We can just use `get_id`. Should we go ahead and do a PR to update that for all the other CRUD classes?